File: Doodream/메뉴 리뉴얼.js
function solution(record) {
    var answer = [];
    var newArr = record.map(str => str.split(" "));

    var nickNameSet = {};
    for (var i = 0; i < newArr.length; i++) {
        if (newArr[i].length === 3) {
            nickNameSet[newArr[i][1]] = newArr[i][2];
        }
    }

    for (var i = 0; i < newArr.length; i++) {
        if (newArr[i][0] === 'Enter') {
            answer.push(nickNameSet[newArr[i][1]] + '님이 들어왔습니다.');
        } else if (newArr[i][0] === 'Leave') {
            answer.push(nickNameSet[newArr[i][1]] + '님이 나갔습니다.');
        }
    }

    return answer;
}

var record = ["Enter uid1234 Muzi", "Enter uid4567 Prodo", "Leave uid1234", "Enter uid1234 Prodo", "Change uid4567 Ryan"];
function solution(orders, course) {
    var answer = [];
    let obj = {}
    orders = orders.map(e => e.split('').sort().join(''));
    for (let i = 0; i < orders.length; i++) {
        let result = [];
        let arr = orders[i].split('');
        for (let j = 0; j < course.length; j++) {
            result.push(...getCombinations(arr, course[j]));
        }
        obj = result.reduce((acc, e) => {
            let joined = e.join('')
            if (joined in acc) {
                acc[joined] += 1;
            } else {
                acc[joined] = 1;
            } return acc
        }, obj)

    }
    let arr = Object.entries(obj);
    for (let i in course) {
        let arr2 = arr.filter(e => e[0].length === course[i] && e[1] !== 1)
        let max = 0
        if (arr2.length > 0) {
            max = Math.max(...arr2.map(e => e[1]))
        }
        arr2.forEach(e => {
            if (e[1] === max) {
                answer.push(e[0])
            }
        })
    }

    return answer.sort();
}

const getCombinations = function (arr, n) {
    const results = [];
    if (n === 1) return arr.map((e) => [e]);
    arr.forEach((e, idx, origin) => {
        const rest = origin.slice(idx + 1);
        const combinations = getCombinations(rest, n - 1);
        const attached = combinations.map((combination) => [e, ...combination]);
        results.push(...attached);
    });
    return results;
}

File: KwangHo/메뉴 리뉴얼.py
from itertools import combinations

def solution(orders, course):
    answer = []
    
    arr=[] #조합넣는곳
    brr={} #조합이 몇개 있는지 개수 저장
    crr=[] #2개이상이면서 최대갯수 인것들 추가
    courseMax={}
    for i in range(len(course)): # course의 최대값 넣을 리스트
        courseMax[course[i]] = 0
    
    for i in range(len(course)):
        for j in range(len(orders)):
            arr = list(map(''.join,combinations(orders[j],course[i]))) #2가지 조합
            for k in range(len(arr)):
                temp = ''.join(sorted(list(arr[k]))) #문자열 정렬후
                if brr.get(temp) == None:
                    brr[temp]=0
                brr[temp]+=1 #참조된 조합 증가
                if courseMax[len(temp)] < brr[temp]: #최대참조 구하기
                    courseMax[len(temp)] = brr[temp]
                
    for i in brr.keys():
        if brr[i] >= courseMax[len(i)] and brr[i]>=2 : #brr이 2이상+최대참조라면 crr에 추가
            crr.append(i)
    crr = sorted(crr) #정렬
    
    answer = crr
    return answer